proxy/src/signal.rs

```diff
@@ -0,0 +1,81 @@
+//! Unix signal handling for the proxy binary.
+
+extern crate futures;
+extern crate tokio_core;
+
+use self::futures::Future;
+use self::tokio_core::reactor::Handle;
+
+type ShutdownSignal = Box<Future<Item=(), Error=()> + Send>;
+
+/// Returns a `Future` that completes when the proxy should start to shutdown.
+pub fn shutdown(handle: &Handle) -> ShutdownSignal {
+    imp::shutdown(handle)
+}
+
+#[cfg(unix)]
+mod imp {
+    extern crate tokio_signal;
+
+    use std::fmt;
+
+    use self::tokio_signal::unix::{Signal, SIGINT, SIGTERM};
+    use super::futures::{future, Future, Stream};
+    use super::{Handle, ShutdownSignal};
+
+    pub(super) fn shutdown(handle: &Handle) -> ShutdownSignal {
+
+        // If you add to this list, you should also update DisplaySignal below
+        // to output nicer signal names.
+        let signals = [SIGINT, SIGTERM]
+            .into_iter()
+            .map(|&sig| {
+                // Create a Future that completes the first
+                // time the process receives 'sig'.
+                Signal::new(sig, handle)
+                    .flatten_stream()
+                    .into_future()
+                    .map(move |_| {
+                        info!(
+                            // use target to remove 'imp' from output
+                            target: "conduit_proxy::signal",
+                            "received {}, starting shutdown",
+                            DisplaySignal(sig),
+                        );
+                    })
+            });
+
+        // With a list of Futures that could notify us,
+        // we just want to know when the first one triggers,
+        // so select over them all.
+
+        let on_any_signal = future::select_all(signals)
+            .map(|_| ())
+            .map_err(|_| unreachable!("Signal never returns an error"));
+
+
+        Box::new(on_any_signal)
+    }
+
+    struct DisplaySignal(i32);
+
+    impl fmt::Display for DisplaySignal {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            let s = match self.0 {
+                SIGINT => "SIGINT",
+                SIGTERM => "SIGTERM",
+                other => return write!(f, "signal {}", other),
+            };
+            f.write_str(s)
+        }
+    }
+}
+
+#[cfg(not(unix))]
```

Comment:
If we're going to use tokio-signal then I'd like for us to follow up with a PR that adds the use of `tokio_signal::ctrl_c` for SIGINT so that we can trigger the same behavior on Windows with Ctrl+C too, since at least two of us develop on Windows. It would be OK to have the `tokio_signal::ctrl_c` implementation only for `not(unix)`.